src/org/sosy_lab/java_smt/test/SolverConcurrencyTest.java

```diff
@@ -0,0 +1,437 @@
+      return factory.generateContext();
+    } catch (InvalidConfigurationException e) {
+      assume()
+          .withMessage(e.getMessage())
+          .that(e)
+          .hasCauseThat()
+          .isNotInstanceOf(UnsatisfiedLinkError.class);
+      throw e;
+    }
+  }
+
+  // TODO: make a Collection of used contextes and end them with annotation after
+  private void closeSolver(SolverContext context) {
+    if (context != null) {
+      context.close();
+    }
+  }
+
+  /**
+   * Synchronizes start and stop of a concurrency test with NUMBER_OF_THREADS threads. This method
+   * will make sure that all exceptions are collected and assessed at the end, as well as stop the
+   * execution after TIMEOUT_SECONDS seconds in case of a deadlock or long calculation.
+   *
+   * @param runnableList A list of Runnable to be tested (make sure its NUMBER_OF_THREADS long for
+   *        proper testing)
+   * @param testName Name of the test for accurate error messages
+   */
+  private static void assertConcurrency(List<? extends Runnable> runnableList, String testName) {
+    int numOfThreads = runnableList.size();
+    final ExecutorService threadPool = Executors.newFixedThreadPool(numOfThreads);
+
+    try {
+      // Waits for all threads to be started
+      final CountDownLatch allExecutorThreadsReady = new CountDownLatch(numOfThreads);
+      // Syncs start of tests after all threads are already created
+      final CountDownLatch afterInitBlocker = new CountDownLatch(1);
+      // Syncs end of tests (And prevents Deadlocks in test-threads etc.)
+      final CountDownLatch allDone = new CountDownLatch(numOfThreads);
+      for (Runnable runnable : runnableList) {
+        threadPool.submit(new Runnable() {
+          @Override
+          public void run() {
+            allExecutorThreadsReady.countDown();
+            try {
+              afterInitBlocker.await();
+              runnable.run();
+            } catch (Throwable e) {
+              exceptionsList.add(e);
+            } finally {
+              allDone.countDown();
+            }
+          }
+        });
+      }
+      assertTrue(
+          "Timeout initializing the Threads for " + testName,
+          allExecutorThreadsReady.await(numOfThreads * 10, TimeUnit.MILLISECONDS));
+      afterInitBlocker.countDown();
+      assertTrue("Timeout in " + testName, allDone.await(TIMEOUT_SECONDS, TimeUnit.SECONDS));
+    } catch (Throwable e) {
+      exceptionsList.add(e);
+    } finally {
+      threadPool.shutdownNow();
+    }
+    assertTrue(
+        "Test " + testName + " failed with exception(s): " + exceptionsList,
+        exceptionsList.isEmpty());
+  }
+
+  /**
+   * As some Solvers are slow... very slow.... and we don't want to wait till the heat death of the
+   * universe nor a 0,1 second test for the faster ones we choose an appropriate number of formulas
+   * to solve here
+   */
+  Map<Solvers, Integer> integerFormulaGen =
```

Comment:
that looks like a constant: `private final Map<Solvers, Integer> INTEGER_FORMULA_GEN`

Reply:
Overall, the tests look fine and already cover the most critical use case, i.e., using different contexts in parallel.

Reply:
Its a constant now.